storage/connect/myutil.cpp

```diff
@@ -182,6 +182,7 @@ int MYSQLtoPLG(int mytype, char *var)
 
   switch (mytype) {
     case MYSQL_TYPE_SHORT:
+    case MYSQL_TYPE_YEAR:
```

Comment:
I would suggest the change happens when the function `MYSQLtoPLG` is called during table creation.
That happens with char arguments. Stack trace is like this:
```gdb
(gdb) bt
+bt
#0  MYSQLtoPLG (typname=0x7ffff159f9d0 "\310\063\001\340\377\177", var=0x1e00348b8 <error: Cannot access memory at address 0x1e00348b8>) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/myutil.cpp:36
#1  0x00005555561fcf2d in MyColumns (g=0x7fffe0023270, thd=0x7fffe0000d28, host=0x7fffd7fff020 "localhost", db=0x7fffe0012f60 "test", user=0x7fffd7fff030 "root", pwd=0x0, table=0x7fffe0012f58 "t1", colpat=0x0, port=16000, info=false) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/myconn.cpp:280
#2  0x00005555561dfc30 in connect_assisted_discovery (thd=0x7fffe0000d28, table_s=0x7ffff15a0fa0, create_info=0x7ffff15a1b50) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/ha_connect.cc:6080
#3  0x0000555555d62b8f in create_table_impl (thd=0x7fffe0000d28, orig_db=0x7fffe0013f70, orig_table_name=0x7fffe0013f80, db=0x7fffe0013f70, table_name=0x7fffe0013f80, path=0x7ffff15a1700 "./test/t2", options={m_options = DDL_options_st::OPT_NONE}, create_info=0x7ffff15a1b50, alter_info=0x7ffff15a1a90, create_table_mode=-3, is_trans=0x7ffff15a195e, key_info=0x7ffff15a16e0, key_count=0x7ffff15a16d4, frm=0x7ffff15a16f0) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_table.cc:5089
#4  0x0000555555d632eb in mysql_create_table_no_lock (thd=0x7fffe0000d28, db=0x7fffe0013f70, table_name=0x7fffe0013f80, create_info=0x7ffff15a1b50, alter_info=0x7ffff15a1a90, is_trans=0x7ffff15a195e, create_table_mode=-3, table_list=0x7fffe0013f58) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_table.cc:5239
#5  0x0000555555d6370f in mysql_create_table (thd=0x7fffe0000d28, create_table=0x7fffe0013f58, create_info=0x7ffff15a1b50, alter_info=0x7ffff15a1a90) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_table.cc:5334
#6  0x0000555555d75e31 in Sql_cmd_create_table_like::execute (this=0x7fffe0013f38, thd=0x7fffe0000d28) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_table.cc:11499
#7  0x0000555555c8a175 in mysql_execute_command (thd=0x7fffe0000d28) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:6076
#8  0x0000555555c8f7b4 in mysql_parse (thd=0x7fffe0000d28, rawbuf=0x7fffe0013da0 "CREATE TABLE t2 ENGINE=CONNECT TABLE_TYPE=MYSQL DBNAME='test' TABNAME='t1' OPTION_LIST='host=localhost,user=root,port=16000'", length=124, parser_state=0x7ffff15a2590, is_com_multi=false, is_next_command=false) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:7855
#9  0x0000555555c7bef1 in dispatch_command (command=COM_QUERY, thd=0x7fffe0000d28, packet=0x7fffe00088d9 "CREATE TABLE t2 ENGINE=CONNECT TABLE_TYPE=MYSQL DBNAME='test' TABNAME='t1' OPTION_LIST='host=localhost,user=root,port=16000'", packet_length=124, is_com_multi=false, is_next_command=false) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:1852
#10 0x0000555555c7a891 in do_command (thd=0x7fffe0000d28) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:1398
#11 0x0000555555dfda52 in do_handle_one_connection (connect=0x555557f94488) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_connect.cc:1403
#12 0x0000555555dfd7ae in handle_one_connection (arg=0x555557f94488) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_connect.cc:1308
#13 0x00005555568dd1c7 in pfs_spawn_thread (arg=0x555557f9bc48) at /home/anel/GitHub/mariadb/server/src/10.3/storage/perfschema/pfs.cc:1869
#14 0x00007ffff77cb609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#15 0x00007ffff76ee133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
During insert and select other implemention (one you used) is called
```gdb
(gdb) bt
+bt
#0  MYSQLtoPLG (mytype=21845, var=0x7ffff15a0ed0 "\220\020Z\361\377\177") at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/myutil.cpp:180
#1  0x00005555561d2280 in ha_connect::GetColumnOption (this=0x7fffe0025b40, g=0x7fffe0027400, field=0x0, pcf=0x7fffd7fff220) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/ha_connect.cc:1656
#2  0x00005555562043e6 in TABDEF::GetColCatInfo (this=0x7fffd7fff070, g=0x7fffe0027400) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/reldef.cpp:456
#3  0x00005555562041ef in TABDEF::Define (this=0x7fffd7fff070, g=0x7fffe0027400, cat=0x7fffe00132f0, name=0x7fffd7fff068 "t2", schema=0x0, am=0x7fffe001a128 "MYSQL") at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/reldef.cpp:379
#4  0x00005555561e9cfb in MYCAT::MakeTableDesc (this=0x7fffe00132f0, g=0x7fffe0027400, tablep=0x7fffd7fff030, am=0x7fffe001a128 "MYSQL") at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/mycat.cc:527
#5  0x00005555561e97f9 in MYCAT::GetTableDesc (this=0x7fffe00132f0, g=0x7fffe0027400, tablep=0x7fffd7fff030, type=0x0) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/mycat.cc:431
#6  0x00005555561ea09b in MYCAT::GetTable (this=0x7fffe00132f0, g=0x7fffe0027400, tablep=0x7fffd7fff030, mode=MODE_READ, type=0x0) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/mycat.cc:549
#7  0x00005555561e6546 in CntGetTDB (g=0x7fffe0027400, name=0x7fffe003043f "t2", mode=MODE_READ, h=0x7fffe0025b40) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/connect.cc:217
#8  0x00005555561d3450 in ha_connect::GetTDB (this=0x7fffe0025b40, g=0x7fffe0027400) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/ha_connect.cc:1997
#9  0x00005555561db32a in ha_connect::info (this=0x7fffe0025b40, flag=18) at /home/anel/GitHub/mariadb/server/src/10.3/storage/connect/ha_connect.cc:4406
#10 0x0000555555db973b in TABLE_LIST::fetch_number_of_rows (this=0x7fffe0013f88) at /home/anel/GitHub/mariadb/server/src/10.3/sql/table.cc:8776
#11 0x0000555555cd4dcc in make_join_statistics (join=0x7fffe0014720, tables_list=@0x7fffe00055b8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7fffe0014df8, last = 0x7fffe0014df8, elements = 1}, <No data fields>}, keyuse_array=0x7fffe0014a10) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_select.cc:4599
#12 0x0000555555ccbbbc in JOIN::optimize_inner (this=0x7fffe0014720) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_select.cc:2000
#13 0x0000555555cca0af in JOIN::optimize (this=0x7fffe0014720) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_select.cc:1542
#14 0x0000555555cd4425 in mysql_select (thd=0x7fffe0000d28, tables=0x7fffe0013f88, wild_num=1, fields=@0x7fffe00054c8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7fffe0013f40, last = 0x7fffe0013f40, elements = 1}, <No data fields>}, conds=0x0, og_num=0, order=0x0, group=0x0, having=0x0, proc_param=0x0, select_options=2147748608, result=0x7fffe00146f8, unit=0x7fffe0004be0, select_lex=0x7fffe00053a0) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_select.cc:4368
#15 0x0000555555cc5658 in handle_select (thd=0x7fffe0000d28, lex=0x7fffe0004b20, result=0x7fffe00146f8, setup_tables_done_option=0) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_select.cc:372
#16 0x0000555555c8b493 in execute_sqlcom_select (thd=0x7fffe0000d28, all_tables=0x7fffe0013f88) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:6340
#17 0x0000555555c81e65 in mysql_execute_command (thd=0x7fffe0000d28) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:3871
#18 0x0000555555c8f7b4 in mysql_parse (thd=0x7fffe0000d28, rawbuf=0x7fffe0013da0 "SELECT * FROM t2", length=16, parser_state=0x7ffff15a2590, is_com_multi=false, is_next_command=false) at /home/anel/GitHub/mariadb/server/src/10.3/sql/sql_parse.cc:7855
```
Maybe we should think of usage of templates or function renaming.
So for now I would recommend to use conversion of `year` to `short` type during creation of table this as an workaround:
```diff
$ git diff
diff --git a/storage/connect/myutil.cpp b/storage/connect/myutil.cpp
index 1c73ded5e05..8287c7539f0 100644
--- a/storage/connect/myutil.cpp
+++ b/storage/connect/myutil.cpp
@@ -40,7 +40,7 @@ int MYSQLtoPLG(char *typname, char *var)
   if (!stricmp(typname, "int") || !stricmp(typname, "mediumint") ||
       !stricmp(typname, "integer"))
     type = TYPE_INT;
-  else if (!stricmp(typname, "smallint"))
+  else if (!stricmp(typname, "smallint") || !stricmp(typname, "year"))
     type = TYPE_SHORT;
   else if (!stricmp(typname, "char") || !stricmp(typname, "varchar") ||
                       !stricmp(typname, "enum") || !stricmp(typname, "set"))
@@ -51,8 +51,7 @@ int MYSQLtoPLG(char *typname, char *var)
   else if (!stricmp(typname, "decimal") || !stricmp(typname, "numeric"))
     type = TYPE_DECIM;
   else if (!stricmp(typname, "date") || !stricmp(typname, "datetime") ||
-           !stricmp(typname, "time") || !stricmp(typname, "timestamp") ||
-           !stricmp(typname, "year"))
+           !stricmp(typname, "time") || !stricmp(typname, "timestamp"))
     type = TYPE_DATE;
   else if (!stricmp(typname, "bigint") || !stricmp(typname, "longlong"))
     type = TYPE_BIGINT;
```

Reply:
![0fc](https://user-images.githubusercontent.com/636271/201306376-d40bec2e-7dff-4b3b-8b98-76061fa218a8.gif)

Good call! I could foresee places where where we might need both my change and your change.

Templating or similar is a good idea, feel free to create a ticket for it, but out of scope of this PR.